associated-token-account/program/src/processor.rs

```diff
@@ -154,3 +160,132 @@ fn process_create_associated_token_account(
+    let nested_token_mint_info = next_account_info(account_info_iter)?;
+    let destination_token_account_info = next_account_info(account_info_iter)?;
+    let destination_system_account_info = next_account_info(account_info_iter)?;
+    let owner_associated_token_account_info = next_account_info(account_info_iter)?;
+    let owner_token_mint_info = next_account_info(account_info_iter)?;
+    let wallet_account_info = next_account_info(account_info_iter)?;
+    let spl_token_program_info = next_account_info(account_info_iter)?;
+    let spl_token_program_id = spl_token_program_info.key;
+
+    // Check owner address derivation
+    let (owner_associated_token_address, bump_seed) =
+        get_associated_token_address_and_bump_seed_internal(
+            wallet_account_info.key,
+            owner_token_mint_info.key,
+            program_id,
+            spl_token_program_id,
+        );
+    if owner_associated_token_address != *owner_associated_token_account_info.key {
+        msg!("Error: Owner associated address does not match seed derivation");
+        return Err(ProgramError::InvalidSeeds);
+    }
+
+    // Check nested address derivation
+    let nested_associated_token_address = get_associated_token_address_with_program_id(
+        owner_associated_token_account_info.key,
+        nested_token_mint_info.key,
+        spl_token_program_id,
+    );
+    if nested_associated_token_address != *nested_associated_token_account_info.key {
+        msg!("Error: Nested associated address does not match seed derivation");
+        return Err(ProgramError::InvalidSeeds);
+    }
+
+    let owner_associated_token_account_signer_seeds: &[&[_]] = &[
+        &wallet_account_info.key.to_bytes(),
+        &spl_token_program_id.to_bytes(),
+        &owner_token_mint_info.key.to_bytes(),
+        &[bump_seed],
+    ];
+
+    if !wallet_account_info.is_signer {
+        msg!("Wallet of the owner associated token account must sign");
+        return Err(ProgramError::MissingRequiredSignature);
+    }
+
+    if owner_token_mint_info.owner != spl_token_program_id {
+        msg!("Owner mint not owned by provided token program");
+        return Err(ProgramError::IllegalOwner);
+    }
+
+    // Account data is dropped at the end of this, so the CPI can succeed
+    // without a double-borrow
+    let (amount, decimals) = {
+        // Check owner associated token account data
+        if owner_associated_token_account_info.owner != spl_token_program_id {
+            msg!("Owner associated token account not owned by provided token program, recreate the owner associated token account first");
+            return Err(ProgramError::IllegalOwner);
+        }
+        let owner_account_data = owner_associated_token_account_info.data.borrow();
+        let owner_account = StateWithExtensions::<Account>::unpack(&owner_account_data)?;
+        if owner_account.base.owner != *wallet_account_info.key {
```

Comment:
nit: so once you setOwner(ata) to another party (i heard that some nft projects do this wrongly?), all of its nested atas are still stuck even with CloseNested. is there any reason for this additional check? how about commenting it here?

Reply:
> is there any reason for this additional check? how about commenting it here?

The main idea was to limit the uses to prevent any possible new vulnerabilities, but it seems safe to remove the check.

The main reason I didn't -- if you `setOwner` your ATA to someone else, who do the nested funds belong to? To the initial owner, or the new owner of the ATA?  Probably the initial owner, right?